apps/opencs/model/world/idtable.cpp

```diff
@@ -281,3 +304,89 @@ CSMWorld::CollectionBase *CSMWorld::IdTable::idCollection() const
 {
     return mIdCollection;
 }
+
+CSMWorld::LandTextureIdTable::LandTextureIdTable(CollectionBase* idCollection, unsigned int features)
+    : IdTable(idCollection, features)
+{
+}
+
+QVariant CSMWorld::LandTextureIdTable::data(const QModelIndex& index, int role) const
+{
+    if (role==Qt::EditRole && !idCollection()->getRecord(index.row()).isModified())
+        return QVariant();
+
+    return IdTable::data(index, role);
+}
+
+bool CSMWorld::LandTextureIdTable::setData(const QModelIndex& index, const QVariant& value, int role)
+{
+    if (!idCollection()->getRecord(index.row()).isModified())
+        return false;
+    else
+        return IdTable::setData(index, value, role);
+}
+
+Qt::ItemFlags CSMWorld::LandTextureIdTable::flags(const QModelIndex& index) const
+{
+    Qt::ItemFlags flags = IdTable::flags(index);
+
+    if (!idCollection()->getRecord(index.row()).isModified())
+        flags &= ~Qt::ItemIsEditable;
+
+    return flags;
+}
+
+CSMWorld::LandTextureIdTable::ImportResults CSMWorld::LandTextureIdTable::importTextures(const std::vector<std::string>& ids)
+{
+    ImportResults results;
+
+    for (const std::string& id : ids)
+    {
+        int plugin, index;
+
+        LandTexture::parseUniqueRecordId(id, plugin, index);
+        int oldRow = idCollection()->searchId(id);
+
+        // If it does not exist or it is in the current plugin, it can be skipped.
+        if (oldRow <= 0 || plugin == 0)
+        {
+            results.recordMapping.push_back(std::make_pair(id, id));
+            continue;
+        }
+
+        // Try a direct mapping to the current plugin first. Otherwise iterate until one is found.
+        // Iteration is deterministic to avoid duplicates.
+        int startIndex = index;
+        do {
+            std::string newId = LandTexture::createUniqueRecordId(0, index);
+            int newRow = idCollection()->searchId(newId);
+
+            if (newRow < 0)
+            {
+                // Id not taken, clone it
+                cloneRecord(id, newId, UniversalId::Type_LandTexture);
+                results.createdRecords.push_back(newId);
+                results.recordMapping.push_back(std::make_pair(id, newId));
+                break;
+            }
+
+            // Id is taken, check if same handle and texture. Note that mId is the handle.
+            const LandTexture& oldLtex = dynamic_cast<const Record<LandTexture>&>(idCollection()->getRecord(oldRow)).get();
+            const LandTexture& newLtex = dynamic_cast<const Record<LandTexture>&>(idCollection()->getRecord(newRow)).get();
+            if (oldLtex.mId == newLtex.mId && oldLtex.mTexture == newLtex.mTexture)
+            {
+                // It's a match
+                results.recordMapping.push_back(std::make_pair(id, newId));
+                break;
+            }
+
+            // Determine next index. Spread out the indices to reduce conflicts.
```

Comment:
We are indeed at risk of running out of indices. However, that limit is 65534. The looping scheme I came up with iterates over all of these indices, though it does not increment how you would traditionally. The gcd of 65534 and 8191 is 1, thus the lcm of the two numbers is 65534 * 8191. Each iteration, 8191 gets added to the index. If you start with an index of 0, then it shouldn't be too difficult to convince yourself that it iterates through all indices.

Reply:
I'm debating if I should construct a map allowing reverse lookups. The benefit is that matching existing LandTexture records will be faster with huge data sets. Also, it would be more accurate for matching LandTexture records from different plugins (an oversight that I just realized). However, I have concerns with keeping the map up to date. It is possible to access the IdCollection directly from the Data class, completely bypassing the IdTable, which could cause some future issues. I suppose the best course of action would be to subclass IdCollection and store the reverse lookup map there. For picking indices, there is no reason to spread out the indices if a reverse lookup map is being used to find matches.

Reply:
Are there performance issues? If not then there is really no reason to introduce error-prone optimisations.

Reply:
Say plugin 1 has an LTEX with grass.dds and index 10 while plugin 2 has an LTEX with grass.dds and index 30. With the current implementation, these will most likely not be matched. I'd say the justification to change it would be for that reason, and not for performance reasons.

Reply:
I'm assuming the above problem I mentioned is not acceptable. 

I think I've come up with a compromise. The importTextures function is only called once for each touch/clone command and is passed all the ids to import for one of these operations. Thus, building a map each time is not entirely wasteful, and more optimal than the alternative which is to iterate over every LTEX record for each passed id. I can test the performance by touching every LAND record, and if it is not satisfactory then we can further discuss this.

Reply:
Sounds reasonable.

Reply:
\o/